### ISSUE TITLE
Support Openshift 3

### ISSUE THREAD

**Issue by yoshiwaan** [fission/fission#107]
Hi,

Great idea and I'm liking the ease of implementation. The company I'm working with at the moment has deployed Openshift 3 rather than pure kubernetes, so I was hoping to have native support for Openshift 3 in Fission.

Openshift is mostly a superset of Kubernetes so it's not not too difficult to get working, I managed to get Fission up and running today with broad strokes, but was hoping to have it done at the application level so there are fewer security concessions.

Openshift has stricter security rules than Kubernetes with a few specifically affecting Fission. That being said the security rules are pretty good, so might be something worth adding to Fission on Kubernetes as well. The main things are:
* Containers run as a random numeric user (so either need a volume or world writable directory to make file system changes)
* Containers run with fewer capabilities
* There is a concept of projects in Openshift. They are directly mapped to namespaces, but you need to explicitly allow permissions for API calls between projects

**Summary**
In short the following needs doing to allow Fission to work on Openshift 3:
1. The fission etcd deployment can't create it's config store directory due to permission. The fix for Openshift would be to specify ETCD_DATA_DIR to either use a volume or create the etcd config dir in something like /tmp, which is world writable
1. The fission controller can't create it's /filestore directory. In Openshift it needs to create the /filestore directory somewhere that is world writable in the container or on a volume. The controller also seem to fail to function correctly unless the filestore directory is created but doesn't panic/error on that failure, so I think it's worth changing the behaviour to panic/exit on /filestore creation failure so that it's easier to debug and more clearly shows as an error (pod restarts)
1. The pool manager doesn't have permission to call cross namespace in Openshift 3, so calls to the fission-function namespace are rejected. The best fix in Openshift would be to create a custom role for the pool manager in the deployment template (with only the API calls to the other project the pool manager needs to make) and assigned it to the correct service account in the fission-function project. An easier but less security fix is to assign the cluster-administrator role in the fission-function project to the default fission service account (system:serviceaccount:fission:default). Due to those missing rules this means that at least the following API calls fail from the pool manager in the current deployment:
   * list extensions.deployments in the fission-function namespace
   * create extensions.deployments in the fission-function namespace
1. It might just be the file permissions above, but until I granted the capability to run as root for the controller (the same set of capabilities as Docker assigns by default) I would get the exception right at the bottom of this ticket. The only way I could fix the issue was to grant those capabilities (as I couldn't change the directory without changing source code like for etcd) so I'm not sure if there's more that needs doing.

**Developing on Openshift**
Openshift is super simple to get up and running, the client tool sets up a docker component based cluster to work with. The CLI tool is also just a superset of Kubectl so there's not much new to learn other than the additional concepts, to get going:
* Get the latest client (cli) tool version from https://github.com/openshift/origin/releases/tag/v1.4.1
* Run `oc cluster up --version 'v1.4.0'`. This will pull down the Openshift Origin containers and run them, making your cluster
* *NOTE:* If using docker toolbox you might need to switch docker-machine context context to the newly created docker machine if it fails half way through and run the same command again, e.g.
  * `eval $(docker-machine env --shell bash openshift)`
  * `oc cluster up --version 'v1.4.0'`
* `oc adm policy add-cluster-role-to-user cluster-admin system --as=system:admin`
* `oc login -u system:admin`
* Use the oc CLI tool (like kubectl) and/or login to the console with the same credentials on port 8443 on the host that is running the docker containers (so maybe your vbox host `docker-machine ip openshift`)

**Getting Fission Deployed on Openshift**
Below are the steps I took to get Fission up and running on Openshift 3, including  a script with the hacks necessary to bypass the security features.

First of all you need to patch http://fission.io/fission.yaml with the following to change the etcd configstore dir:
```
142a143,144
>         - name: ETCD_DATA_DIR
>           value: /tmp/default.etcd
```

Then running the following script will do the job (in a shell that is logged in with `oc`):
```
#!/bin/bash

 oc new-project fission # The existing fission templates already create the project correctly by using the namespace attribute, but I still do it in order to make the changes below before creation of the resources
 oc adm policy add-scc-to-user anyuid -z default # Hack 1
 oc new-project fission-function
 oc policy add-role-to-user cluster-admin system:serviceaccount:fission:default # Hack 2
 oc create -f fission.yaml # Hack 3
 oc create -f fission-nodeport.yaml
```

1. Hack 1: This fixes item number 2 from the summary section, but in a broad way which gives the container more capabilities than it needs
1. Hack 2: This fixes item number 3 from the summary section, but in a way that is overly permissive for the pool manager.
1. Hack 3: Deploying the etcd pod so it creates the configstore in /tmp, not necessarily a hack.

With the above hacks I was able to successfully deploy the hello world function, so it's definitely possible to get it working!

**Exception**
The exception referenced in the last point of the summary section
```
2017/02/06 00:45:07 http: panic serving 172.17.0.1:59594: runtime error: invalid memory address or nil pointer dereference
goroutine 4036 [running]:
net/http.(*conn).serve.func1(0x19f62e00)
	/usr/local/go/src/net/http/server.go:1491 +0xa2
panic(0x8fe7940, 0x19c1e008)
	/usr/local/go/src/runtime/panic.go:458 +0x40b
github.com/fission/fission/controller.(*ResourceStore).writeFile(0x19e2bea0, 0x1a018a80, 0x5, 0x1a073800, 0x53, 0x60, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/soam/gopath/src/github.com/fission/fission/controller/resourceStore.go:156 +0x18d
github.com/fission/fission/controller.(*FunctionStore).Create(0x19e2bea0, 0x19f3c9c0, 0x0, 0x0, 0x0, 0x0)
	/Users/soam/gopath/src/github.com/fission/fission/controller/functionStore.go:31 +0xec
github.com/fission/fission/controller.(*API).FunctionApiCreate(0x19e2bea0, 0xf76a6988, 0x19f13fc0, 0x1a032380)
	/Users/soam/gopath/src/github.com/fission/fission/controller/functionApi.go:67 +0x232
github.com/fission/fission/controller.(*API).FunctionApiCreate-fm(0xf76a6988, 0x19f13fc0, 0x1a032380)
	/Users/soam/gopath/src/github.com/fission/fission/controller/api.go:73 +0x38
net/http.HandlerFunc.ServeHTTP(0x19fdf738, 0xf76a6988, 0x19f13fc0, 0x1a032380)
	/usr/local/go/src/net/http/server.go:1726 +0x34
github.com/fission/fission/vendor/github.com/gorilla/mux.(*Router).ServeHTTP(0x19e4c270, 0xf76a6988, 0x19f13fc0, 0x1a032380)
	/Users/soam/gopath/src/github.com/fission/fission/vendor/github.com/gorilla/mux/mux.go:103 +0x210
github.com/fission/fission/vendor/github.com/gorilla/handlers.loggingHandler.ServeHTTP(0x9887b70, 0x19c1e0d8, 0x9885430, 0x19e4c270, 0x988dd40, 0x19d041b0, 0x1a032380)
	/Users/soam/gopath/src/github.com/fission/fission/vendor/github.com/gorilla/handlers/handlers.go:69 +0xf0
github.com/fission/fission/vendor/github.com/gorilla/handlers.(*loggingHandler).ServeHTTP(0x1a031240, 0x988dd40, 0x19d041b0, 0x1a032380)
	<autogenerated>:8 +0x9b
net/http.serverHandler.ServeHTTP(0x19f88410, 0x988dd40, 0x19d041b0, 0x1a032380)
	/usr/local/go/src/net/http/server.go:2202 +0x136
net/http.(*conn).serve(0x19f62e00, 0x988e940, 0x19f13e40)
	/usr/local/go/src/net/http/server.go:1579 +0xe86
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:2293 +0x489
```

**Comment by soamvasani** [comment]
Thanks @yoshiwaan, your detailed investigation is very very helpful!  Great to know you got it working on openshift!

> The fission etcd deployment can't create it's config store directory due to permission. The fix for Openshift would be to specify ETCD_DATA_DIR to either use a volume or create the etcd config dir in something like /tmp, which is world writable

I wonder if the etcd operator helps with this stuff.  If we install via helm chart (pr #90) then perhaps we can just depend on the etcd operator for this, and not have to solve it in fission.

> changing the behaviour to panic/exit on /filestore creation failure so that it's easier to debug and more clearly shows as an error (pod restarts)

I've filed #108 to track this bug.

>  The best fix in Openshift would be to create a custom role for the pool manager in the deployment template (with only the API calls to the other project the pool manager needs to make) and assigned it to the correct service account in the fission-function project. An easier but less security fix is to assign the cluster-administrator role in the fission-function project to the default fission service account 

Do you know if there any way to do this in the YAML file?  Then we could have a fission-openshift.yaml that does either one of these things.

> exception

This stack trace looks like the error handling bug #108.  The fix for the root cause is probably to provide a volume at /filestore (or some other path, with the path specified in the controller args).

**Comment by yoshiwaan** [comment]
I've not used etcd-operator so I can't comment unfortunately, but I'm surmising that you mean for users to just specify their own config files which specify ETCD_DATA_DIR in that case?
Do you have a rough timeline for integrating helm? The short term fix of updating your template could do in the meantime to work on both platforms.

I spent a bit of time looking about how to provide the correct security permissions for fission on Openshift and came up with this:
```
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: poolmgr
  namespace: fission
---
apiVersion: v1
kind: ClusterRole
metadata:
  name: fission:poolmgr
rules:
- apiGroups:
  - extensions
  attributeRestrictions: null
  resources:
  - deployments
  verbs:
  - create
  - get
  - list
  - update
  - watch
- apiGroups:
  - ""
  attributeRestrictions: null
  resources:
  - pods
  verbs:
  - get
  - list
  - update
---
apiVersion: v1
groupNames: null
kind: RoleBinding
metadata:
  name: fission:poolmgr
  namespace: fission-function
roleRef:
  name: fission:poolmgr
subjects:
- kind: ServiceAccount
  name: poolmgr
  namespace: fission
userNames:
- system:serviceaccount:fission:poolmgr
```
Adding the above at the top of the fission.yaml template (under the namespace creation) will create a service account for the poolmgr in the fission namespace and a role at the cluster level which allows just the API calls needed (that I have found so far) for the poolmgr to execute the API commands it needs.

The rolebinding then binds the serviceaccount that the poolmgr runs to the new role.

The final piece is to alter the poolmgr deployment to run as the new service account rather than the default for the namespace:
```
---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: poolmgr
  namespace: fission
spec:
  replicas: 1
  template:
    metadata:
      labels:
        svc: poolmgr
    spec:
      containers:
      - name: poolmgr
        image: fission/fission-bundle:testing20170124
        command: ["/fission-bundle"]
        args: ["--poolmgrPort", "8888"]
      serviceAccount: poolmgr
```

If you need to add more API calls for the poolmgr then you can just update the role.

Note that this is all Kubernetes stuff as well (in the newer versions?) but is only adhered to in certain permissions mode (my understanding is that Openshift runs in the RBAC permissions mode). I believe you can create all of these objects in vanilla k8s.

The Security Context Constraint (SCC) looks like it's Openshift specific at the moment (there is some documentation on Security Contexts in k8s but it doesn't look as fleshed out as the Openshift version) and I can't find an object you can create to replicate the add-scc-to-user command I gave. There might be one out there, but the better fix is to update your code to work with a numeric random UID anyway.

**Comment by Vad1mo** [comment]
I am trying to run fission on openshift. Following @yoshiwaan's [explanation](https://github.com/fission/fission/issues/107#issue-205475310) ending up with 5 services running. I created and external route to the router. After following the [example](https://github.com/fission/fission#run-an-example) I ended up doing a  `curl http://$FISSION_ROUTER/hello` but there I only get the 503 response `Internal server error (fission)` 

```
2017/03/21 20:42:42 Not cached, getting new service for {hello 3040a006-513b-43f8-8e8d-b26f676274d0}
192.168.16.1 - - [21/Mar/2017:20:42:42 +0000] "GET /hello HTTP/1.1" 500 32
2017/03/21 20:42:42 Failed to get service for function (hello,3040a006-513b-43f8-8e8d-b26f676274d0): (Error 0) 500 Internal Server Error
```

```
$fission route list
NAME                                 METHOD URL    FUNCTION_NAME FUNCTION_UID
e31d850c-6f53-442e-9dd0-3456e39e417c GET    /hello hello 

$fission fn list
NAME  UID                                  ENV
hello 3040a006-513b-43f8-8e8d-b26f676274d0 nodejs

$fission fn getmeta hello
Failed to get function: (Error 2) HTTP error 404
$fission fn getmeta 3040a006-513b-43f8-8e8d-b26f676274d0
Failed to get function: (Error 2) HTTP error 404
```


So I am wondering why its not working as expected in the example

**Comment by methadata** [comment]
I have been able to code in a single yaml (deployment config file) all the needed changes to deploy Fission on OpenShift, using the great feedback provided by @yoshiwaan 

The main addition/difference in #168 from previous comments in this issue is the use of an *emptyDir* volume for the controller that is mounted in `/filestore`. This way, there's no need to add the default user to a new Security Context (SCC), and thus not using `anyuid`, as used in *Hack 1*. This avoids the exception in the controller container.

Hack 2, 3 and security permissions have been incorporated in fission-openshift.yaml file.

I have tested the deployment with teh following versions:
oc v1.4.1+3f9807a
kubernetes v1.4.0+776c994
openshift v1.4.1+3f9807a

Also, it would be great to try to modify the deployment in order to use serviceAccounts with regular roles, instead of those having a ClusterRole. If so, the whole deployment could be launched by any project user, and not a cluster-admin (system:admin).

Hope this helps!

**Comment by methadata** [comment]
I would say so. Nevertheless, it's a long time since I haven't tested the latest release depoyment in OpenShift.

**Comment by faust64** [comment]
FYI, just trying to deploy Fission (1.6.0) on OpenShift (4), and half my Pods are crashed:

```
$ oc new-project fission
$ kubectl -n fission apply -f https://github.com/fission/fission/releases/download/1.6.0/fission-all-1.6.0-minikube.yaml
...
$ oc get pods
NAME                                                           READY   STATUS                  RESTARTS   AGE
buildermgr-595d55b8c8-8rql7                                    1/1     Running                 0          6m20s
controller-844f5ddbf6-7l6xv                                    1/1     Running                 0          6m21s
executor-5d4bfdfdb5-r6srk                                      1/1     Running                 0          6m20s
fission-1-6-0-fission-al-1.6.0-125-8fqvg                       0/1     OOMKilled               0          6m6s
fission-1-6-0-fission-al-1.6.0-183-l62xh                       0/1     Error                   0          6m5s
fission-1-6-0-prometheus-alertmanager-6859c5b46-gzftd          2/2     Running                 0          6m9s
fission-1-6-0-prometheus-kube-state-metrics-7dbb5fcfc9-7rr8j   1/1     Running                 0          6m9s
fission-1-6-0-prometheus-pushgateway-5c955d69b8-czvpm          1/1     Running                 0          6m8s
fission-1-6-0-prometheus-server-85c765b6c9-rlcnd               0/2     Init:CrashLoopBackOff   5          6m7s
influxdb-55bdd84774-cnb4j                                      1/1     Running                 0          6m19s
kubewatcher-54667f9cc6-tld5c                                   1/1     Running                 1          6m20s
logger-2t5bs                                                   0/2     Init:Error              6          6m16s
logger-56rw2                                                   0/2     Init:Error              6          6m15s
logger-5lbtc                                                   0/2     Init:CrashLoopBackOff   6          6m16s
logger-8884d                                                   0/2     Init:Error              6          6m16s
logger-d68wp                                                   0/2     Init:CrashLoopBackOff   6          6m16s
logger-pdhcp                                                   0/2     Init:CrashLoopBackOff   6          6m16s
logger-v7tj6                                                   0/2     Init:CrashLoopBackOff   6          6m16s
logger-x64dz                                                   0/2     Init:Error              6          6m16s
mqtrigger-nats-streaming-6df77d5dd8-cs8lj                      1/1     Running                 6          6m18s
redis-0                                                        1/1     Running                 0          6m11s
router-5dc849d7dc-gt4wq                                        1/1     Running                 0          6m7s
storagesvc-56447cd957-mn8c4                                    1/1     Running                 0          6m18s
timer-76b97ff487-75vjn                                         1/1     Running                 0          6m19s
```

The logger containers can't write /var/log/fission:

```
$ oc logs logger-5lbtc -c init
mkdir: can't create directory '/var/log/fission': Permission denied
```

The Prometheus init container can't change permissions on /data:

```
$ oc logs fission-1-6-0-prometheus-server-85c765b6c9-rlcnd -c init-chown-data
chown: /data: Operation not permitted
chown: /data: Operation not permitted
```

Whereas the fission container seems to have issues querying objects against OpenShift API:

```
$ oc logs fission-1-6-0-fission-al-1.6.0-183-l62xh
{"level":"info","ts":1573574481.3234055,"logger":"pre_upgrade_task_client","caller":"preupgradechecks/preupgradechecks.go:87","msg":"verifying function spec references for all functions in the cluster"}
{"level":"fatal","ts":1573574481.3501542,"logger":"pre_upgrade_task_client","caller":"preupgradechecks/preupgradechecks.go:128","msg":"installation failed","summary":"a function cannot reference secrets, configmaps and packages outside it's own namespace","stacktrace":"main.(*PreUpgradeTaskClient).VerifyFunctionSpecReferences\n\tsrc/cmd/preupgradechecks/preupgradechecks.go:128\nmain.main\n\tsrc/cmd/preupgradechecks/main.go:69\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:200"}
```

Although OpenShift 4 is slightly different from OpenShift 3, those errors don't look anything new.

FYI, I've been following the "without helm", "all-on-minikube" section from https://docs.fission.io/docs/installation/

**Comment by life1347** [comment]
@faust64 Thanks for providing such useful troubleshooting information. 

> $ oc logs logger-5lbtc -c init
> mkdir: can't create directory '/var/log/fission': Permission denied

So looks like openshift doesn't allow to create directory under `/var/log`, it's easy to fix. 

> $ oc logs fission-1-6-0-prometheus-server-85c765b6c9-rlcnd -c init-chown-data
> chown: /data: Operation not permitted

Prometheus is the optional chart dependency for fission. So this is out of scope we can control. However, you can check the NOTE at this [page](https://wiki.berki.org/index.php/Prometheus_stack_on_Kubernetes#Prometheus_stack_install_with_Helm). It should help you to solve the problem.

> $ oc logs fission-1-6-0-fission-al-1.6.0-183-l62xh
>{"level":"info","ts":1573574481.3234055,"logger":"pre_upgrade_task_client","caller":"preupgra..

It's a bug in preupgradecheck, I've fixed and will be shipped in 1.7.0, which will come out in next couple days.

If you don't mind, please join the Slack channel and ping me.
